federatedscope/core/auxiliaries/parallel_runner.py

```diff
@@ -0,0 +1,447 @@
+                              data[1],
+                              backend=self.config.backend)
+        kw = {
+            'shared_comm_queue': self.send_channel,
+            'id2comm': self.id2comm,
+            'resource_info': self.resource_info,
+            'client_resource_info': self.client_resource_info
+        }
+
+        self.server = self.server_class(
+            ID=self.server_id,
+            config=self.config,
+            data=self.data,
+            model=model,
+            client_num=self.config.federate.client_num,
+            totol_round_num=self.config.federate.total_round_num,
+            device=self.device,
+            unseen_clients_id=self.unseen_clients_id,
+            **kw)
+
+        self.server.model.to(self.device)
+        self.template_para = copy.deepcopy(self.server.model.state_dict())
+        if self.config.nbafl.use:
+            from federatedscope.core.trainers.trainer_nbafl import \
+                wrap_nbafl_server
+            wrap_nbafl_server(self.server)
+        logger.info('Server has been set up ... ')
+        _, feat_engr_wrapper_server = get_feat_engr_wrapper(self.config)
+        self.server = feat_engr_wrapper_server(self.server)
+
+    def run(self):
+        logger.info("ServerRunner {} start to run".format(self.rank))
+        server_msg_cache = list()
+        while True:
+            if not self.receive_channel.empty():
+                msg = self.receive_channel.get()
+                # For the server, move the received message to a
+                # cache for reordering the messages according to
+                # the timestamps
+                heapq.heappush(server_msg_cache, msg)
+            elif len(server_msg_cache) > 0:
+                msg = heapq.heappop(server_msg_cache)
+                if self.config.asyn.use and self.config.asyn.aggregator \
+                        == 'time_up':
+                    # When the timestamp of the received message beyond
+                    # the deadline for the currency round, trigger the
+                    # time up event first and push the message back to
+                    # the cache
+                    if self.server.trigger_for_time_up(msg.timestamp):
+                        heapq.heappush(server_msg_cache, msg)
+                    else:
+                        self._handle_msg(msg)
+                else:
+                    self._handle_msg(msg)
+            else:
+                if self.config.asyn.use and self.config.asyn.aggregator \
+                        == 'time_up':
+                    self.server.trigger_for_time_up()
+                    if self.client2server_comm_queue.empty() and \
+                            len(server_msg_cache) == 0:
+                        break
+                    else:
+                        # terminate when shared_comm_queue and
+                        # server_msg_cache are all empty
+                        time.sleep(0.01)
```

Comment:
We need to terminate the FL course when finished, maybe we can use `server.is_finish`?